worker/task/process_event_channel.go

```diff
@@ -0,0 +1,238 @@
+	MatchSubscriptions(context.Context, EventChannelMetadata, datastore.EventRepository, datastore.ProjectRepository, datastore.EndpointRepository, datastore.SubscriptionRepository, license.Licenser) (*EventChannelSubResponse, error)
+}
+
+func ProcessEventCreationByChannel(channel EventChannel, endpointRepo datastore.EndpointRepository, eventRepo datastore.EventRepository, projectRepo datastore.ProjectRepository, eventQueue queue.Queuer, subRepo datastore.SubscriptionRepository, licenser license.Licenser) func(context.Context, *asynq.Task) error {
+	return func(ctx context.Context, t *asynq.Task) error {
+		cfg := channel.GetConfig()
+
+		// get or create event
+		var lastEvent, lastRunErrored, err = getLastTaskInfo(ctx, t, channel, eventQueue, eventRepo)
+		if lastEvent != nil && lastEvent.IsDuplicateEvent && !lastRunErrored {
+			log.FromContext(ctx).Debugf("[asynq]: duplicate event with idempotency key %v will not be sent", lastEvent.IdempotencyKey)
+			return nil
+		}
+
+		var event *datastore.Event
+		if lastEvent != nil {
+			log.Debug("[asynq] processing last event")
+			event = lastEvent
+		} else {
+			log.Debug("[asyq] creating new event")
+			event, err = channel.CreateEvent(ctx, t, channel, eventRepo, projectRepo, endpointRepo, subRepo, licenser)
+			if err != nil {
+				if strings.Contains(err.Error(), "duplicate key") {
+					lastEvent, err = eventRepo.FindEventByID(ctx, event.ProjectID, event.UID)
+				}
+
+				if lastEvent == nil {
+					writeErr := fmt.Errorf("failed to create event, err: %s", err.Error())
+					err = &EndpointError{Err: writeErr, delay: cfg.DefaultDelay}
+					return err
+				}
+
+				log.WithError(err).Error("skipping duplicated event: " + event.UID)
+				return nil
+			}
+			if event == nil {
+				return &EndpointError{Err: fmt.Errorf("CODE: 1009, no response, failed to create event via channel %s", cfg.Channel), delay: cfg.DefaultDelay}
+			}
+		}
+
+		metadata := EventChannelMetadata{
+			Event:  event,
+			Config: cfg,
+		}
+
+		payload, err := msgpack.EncodeMsgPack(metadata)
+		if err != nil {
+			return err
+		}
+
+		jobId := fmt.Sprintf("match_subs:%s:%s", event.ProjectID, event.UID)
+		job := &queue.Job{
+			ID:      jobId,
+			Payload: payload,
+			Delay:   0,
+		}
+
+		log.Info("[asynq] writing to queue to match subscriptions")
+		err = eventQueue.Write(convoy.MatchEventSubscriptionsProcessor, convoy.EventWorkflowQueue, job)
+		if err != nil {
+			log.FromContext(ctx).WithError(err).Errorf("[asynq]: an error occurred while matching event subs")
+		}
+
+		return err
+	}
+}
+
+func MatchSubscriptionsAndCreateEventDeliveries(channels map[string]EventChannel, endpointRepo datastore.EndpointRepository, eventRepo datastore.EventRepository, projectRepo datastore.ProjectRepository, eventDeliveryRepo datastore.EventDeliveryRepository, eventQueue queue.Queuer, subRepo datastore.SubscriptionRepository, deviceRepo datastore.DeviceRepository, licenser license.Licenser) func(context.Context, *asynq.Task) error {
```

Comment:
All these plenty params, can we create a struct to hold them? This goes for all the other methods with a lot of params as well.

Reply:
Noted.